Data format descriptor

superdump opened the pull request:
## Checklist

- [x] `cargo clippy` reports no issues
- [x] `cargo doc` reports no issues
- [x] [`cargo deny`](https://github.com/EmbarkStudios/cargo-deny/) issues have been fixed or added to `deny.toml`
- [x] `cargo test` shows all tests passing
- [x] human-readable change descriptions added to the changelog under the "Unreleased" heading.
  - [x] If the change does not affect the user (or is a process change), preface the change with "Internal:"
  - [x] Add credit to yourself for each change: `Added new functionality @githubname`.

## Description

Add support for parsing Data Format Descriptor sections from KTX2 files.

superdump committed:
Initial shot at parsing the Data Format Descriptor sections

superdump committed:
Appease clippy

This is functionally the same but perhaps an Unsupported variant should be
added to the enums for the catch all?

superdump committed:
Add credit to changelog

superdump commented:
To appease clippy I had to change some `0 | _ => SomeType::Unspecified` to just `_ =>` as it rightly points out that the `_` matches anything else anyway. I had left it as `0 | _` as some kind of documentation. `0` is explicitly in the spec as an 'unspecified' type, presumably so it can be used for custom data without breaking things for other people as ids get standardised. However, as ids get standardised, they won't be in the enums. It feels wrong that a constructor function would return a `Result` so perhaps adding an `Unsupported` to the enums and mapping `0` to `Unspecified` as per the spec, and anything that the crate doesn't know about to `Unsupported`. However, it would be a shame to break/block user code just because it is an unknown type by the crate. Maybe instead adding an `Unknown(pub u32)` or whatever is a better option...? Mmm, I like that last option.

superdump committed:
Add Unknown(u32) for standardized enum values unknown by the crate

Ralith commented on src/lib.rs:
```diff
@@ -98,6 +99,35 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
+        self.input.as_ref()[start..end].try_into().unwrap()
+    }
+
+    pub fn data_format_descriptors(&self) -> Vec<BasicDataFormatDescriptor> {
```
Let's make this a non-allocating `Iterator` to keep the crate alloc-free. The caller can always `collect` if they want.

superdump replied:
Ok, I’ll look into that.

superdump replied:
Done. I also implemented a similar iterator for the SampleInformation blocks so that I could get rid of Vec entirely. However, it was a little more awkward because while `data_format_descriptors()` is on the top-level reader, it felt like the `sample_informations()` should be on the individual `BasicDataFormatDescriptor` structs. But those don't have the underlying file data. So, I opted to pass it into the `sample_informations(data: &[u8])`. I did try a different way first which had the BasicDataFormatDescriptor hold a borrow on the underlying data for its entire block, but that meant adding an explicit lifetime parameter to the struct which felt ugly to me.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +208,371 @@ impl LevelIndex {
+}
+
+#[derive(Debug)]
+pub enum ColorModel {
```
Extensible enums should use the `pseudo_enum` macro from enum.rs (and perhaps be defined there as well). This allows new values to be added without breaking compatibility.

superdump replied:
Interesting, I’ll take a look at that.

superdump replied:
I moved the color model, primaries, and transfer function to enums.rs and used the pseudo_enum macro. However, because it uses NonZeroU32, it cannot represent the 'unspecified' value from the spec that has a value of 0. I chose to add these members of the `BasicDataFormatDescriptor` as `Option<T>` and have `None` mean unspecified. Hopefully that is acceptable. If not, I'm open to suggestions.

Ralith replied:
That is the intended pattern, thanks!

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +208,371 @@ impl LevelIndex {
+    }
+}
+
+#[inline]
```
Unless profiling has clearly indicated a benefit, let's avoid `#[inline]` on internal helper functions; LLVM's heuristics can generally handle those.

superdump replied:
I’ll remove them.

superdump replied:
Done.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +208,371 @@ impl LevelIndex {
+        const EXPONENT      = (1 << 1);
+        const SIGNED        = (1 << 2);
+        const FLOAT         = (1 << 3);
+        const NONE          = 0;
+        const UNINITIALIZED = 0xFFFF;
```
What's the purpose of these constants? They don't seem to be present in the DFD spec, afaict.

superdump replied:
No indeed they aren’t in the spec. They’re a pattern from somewhere and I’ve copied, pasted, and modified. I’m yet to find a use for uninitialised, but none is useful, albeit redundant. I’ll remove them.

superdump replied:
Removed.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +208,371 @@ impl LevelIndex {
+bitflags::bitflags! {
+    #[derive(Default)]
+    #[repr(transparent)]
+    pub struct DataFormatFlags: u32 {
```
If this bitfield is subject to future extension, bitflags may not be a good choice for it, as some if its operators (e.g. `all`) don't make sense if there may be valid bits unknown to the code.

superdump replied:
I can implement something manually if you prefer.

Ralith replied:
I'd lean towards that, maybe with a helper macro for the boilerplate. We made a similar mistake in ash and rolled it back in the end.

superdump replied:
Hmm, I'm basically going to have to reimplement the bitflags crate. I don't want to just copy and paste it into this crate just to remove the `all()` trait method, nor do I want to reimplement its functionality. Plus I'm not yet fluent with macros which means it will be pretty slow going for me.

Ralith replied:
This isn't super important to me, just a minor nit. @cwfitzgerald, thoughts?

cwfitzgerald replied:
I think leaving it as bitflags is fine, all isn't a query that is particularly relevant to this flag set, so it shouldn't be a big footgun.

Ralith reviewed (changes requested):
Thanks for working on this! Out of curiosity, what's your use case for this data?

superdump commented:
> Thanks for working on this! Out of curiosity, what's your use case for this data?

tl;dr so I can load supercompressed and compressed textures from KTX2 files.

I pulled in the `ktx2` crate to bevy to try to add support for some compressed texture formats (also `ddsfile`). As I was hooking things up and trying to get something working, I noticed that the sample I had was supercompressed and had no format. I looked at the spec and it seemed like parsing the Data Format Descriptor was necessary to figure out the format of the contained data. Indeed after implementing this I was able to understand that the textures contained in the sample that I had were UASTC format, which I think need additional transcoding to formats that can be directly sampled by a GPU like BC*/ASTC/ETC2.

So, my goal is support for KTX2 as a container primarily for compressed textures, with or without mipmaps, and perhaps also supporting cubemaps, even if long-term bevy has a different and custom asset metadata format to be able to configure whatever the engine needs.

The asset pipeline is the next priority in bevy I think, but it’s good to have good support for glTF 2.0, and KTX2/DDS out of the box.

superdump committed:
Remove explicit #[inline]

superdump committed:
Remove UNINITIALIZED and NONE flags

superdump committed:
Make data_format_descriptors() return an Iterator

Ralith commented:
> I noticed that the sample I had was supercompressed and had no format

Interesting, didn't realize that was at all common. That makes this a pretty key feature!

superdump committed:
Add DataFormatDesctiptor::sample_informations(data: &[u8]) that returns an Iterator

superdump committed:
Use pseudo_enum for color model, primaries, and transfer function enums

Note that while 0 means Unspecified, and ironically is defined in the
specification with a particular meaning, because pseudo_enum uses NonZeroU32,
it cannot represent 0. As such, I chose to make the members Option<T> and have
None mean Unspecified.

superdump committed:
cargo fmt

Ralith commented on src/lib.rs:
```diff
@@ -325,8 +313,35 @@ impl BasicDataFormatDescriptor {
+        }
+    }
+
+    pub fn sample_informations<'data>(&self, data: &'data [u8]) -> impl Iterator<Item = SampleInformation> + 'data {
```
Rather than requiring the user to pass in the correct `data`, let's pipe it through `self`. If the lifetime parameter becomes an issue in practice (which I don't expect) we can split out a separate struct for just the fixed-size parts. 

Additionally, "information" is its own plural, so let's drop the trailing s from this function name.

superdump replied:
> Rather than requiring the user to pass in the correct `data`, let's pipe it through `self`. If the lifetime parameter becomes an issue in practice (which I don't expect) we can split out a separate struct for just the fixed-size parts.

Ok. I was thinking that the lifetime being introduced to the API is not really a problem as you anyway have the data for the file available throughout usage of it. I'll do that.

> Additionally, "information" is its own plural, so let's drop the trailing s from this function name.

Yup.

superdump replied:
Done.

superdump committed:
Pipe data through BasicDataFormatDescriptor for SampleInformationIterator

Ralith commented on src/lib.rs:
```diff
@@ -98,6 +96,49 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
+        let header = self.header();
+        let start = header.sgd_byte_offset as usize;
+        let end = (header.sgd_byte_offset + header.sgd_byte_length) as usize;
+        self.input.as_ref()[start..end].try_into().unwrap()
```
What exactly is `.try_into().unwrap()` doing here?

Ralith commented on src/lib.rs:
```diff
@@ -98,6 +96,49 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
+        let header = self.header();
+        let start = header.dfd_byte_offset as usize;
+        let length: u32 = u32::from_le_bytes(self.input.as_ref()[start..start + 4].try_into().unwrap());
+        assert_eq!(length, header.dfd_byte_length);
+        assert_eq!(length, header.kvd_byte_offset - header.dfd_byte_offset);
```
Are these asserts defending against malformed input data or bugs in this library? If the former, we should try to check during `new` instead. If the latter, switch to `debug_assert`.

superdump replied:
I'll move the checks to new. Should be to check the validity of input data.

superdump replied:
Although, should we deny parsing a file at all if only part of it is corrupt?

Ralith replied:
I think that's the correct default, and helps prevent the widespread dissemation of invalid data to begin with. It also makes the user's life easier, because they can rely on a `ktx2::Reader` always representing valid data, rather than having to validate every possible malformed component themselves and/or litter their code with `unwrap`s.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+}
+
+impl<'data> BasicDataFormatDescriptor<'data> {
+    pub fn from_bytes_and_offset(bytes: &'data [u8], initial_offset: usize) -> Self {
```
Will downstream code ever need to call this directly? If not, it shouldn't be `pub`.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+impl SampleInformation {
+    const LENGTH: usize = 16;
+
+    pub fn from_bytes(bytes: &[u8]) -> Self {
```
As above.

superdump committed:
Address some review feedback

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+// This can be obtained from std::mem::transmute::<f32, u32>(1.0f32) but we
+// want to support nostd. It is used for identifying normalized sample types
+// as in Unorm or Snorm
+const F32_1_AS_U32: u32 = 1065353216;
```
Could [`core::mem::transmute`](https://doc.rust-lang.org/core/mem/fn.transmute.html) be used rather than hardcoding a magic number?

superdump replied:
This will be removed as noted below.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+        }
+    }
+
+    pub fn is_exponent(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::EXPONENT)
+    }
+
+    pub fn is_float(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::FLOAT)
+    }
+
+    pub fn is_linear(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::LINEAR)
+    }
+
+    pub fn is_signed(&self) -> bool {
+        self.channel_type_qualifiers.contains(ChannelTypeQualifiers::SIGNED)
+    }
+
+    pub fn is_norm(&self) -> bool {
+        self.is_float() && self.upper == F32_1_AS_U32
+    }
```
Do these offer significant value to callers, given that the fields and enums involved are already public and (presumably) correspond more directly to the spec?

superdump replied:
I'll just do it where I've been calling these. And so I'll remove the `F32_1_AS_U32` constant.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+    pub channel_type: u32,                              //: 4;
+    pub channel_type_qualifiers: ChannelTypeQualifiers, //: 4;
+    pub sample_positions: [u32; 4],                     //: 8 x 4;
+    pub lower: u32,                                     //;
+    pub upper: u32,                                     //;
```
From the use in `is_norm` I gather these values have structure we're not capturing. Is there a way we could abstract over them more ergonomically?

superdump replied:
The is_norm is my own invention based on what I saw in files. The bounds can be anything anyone wants, but they will commonly be signed/unsigned int / unorm/snorm/float. But this is kind of specific I guess so probably better to keep this code outside the crate where the usage of it is known.

Ralith replied:
Ah, I had guessed these were "lower/upper 32 bits", not "lower/upper bound". That makes more sense. A glance at the spec suggests that the additional structure here is quite complex so passing through the bits as done here is probably the right place for us to start.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+pub struct BasicDataFormatDescriptor<'data> {
+    pub vendor_id: u32,             //: 17;
+    pub descriptor_type: u32,       //: 15;
+    pub version_number: u32,        //: 16;
```
Is it possible for `version_number` to vary without changing the semantics/presence of other fields? If not, we might as well omit it.

superdump replied:
The Data Format Descriptor blocks always have the same first 8 bytes containing the vendor id, descriptor type, version number, and descriptor block size: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_anchor_id_descriptorblock_xreflabel_descriptorblock_descriptor_block . These are to allow for extension with many different types of data format descriptor to express whatever information someone may need.

Khronos primarily define the Basic Data Format Descriptor that I have implemented parsing for which covers all common compressed and uncompressed texture formats and honestly already much more: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_anchor_id_basicdescriptor_xreflabel_basicdescriptor_khronos_basic_data_format_descriptor_block

Khronos also define two 'advanced' data format descriptors that are extensions to the supported number of dimensions and their corresponding byte planes. For example, the format may have 'planes' due to subsampling of channel data. Y'CbCr 4:2:0 is given as an example where it has dimensions 2x2x1x1 (the basic data format descriptor covers four dimensional textures by default) and the Y' samples are split into two planes which are offset by one line in the y dimension. The data format descriptor for additional byte planes: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_additional_planes_descriptor_block and for additional dimensions: https://www.khronos.org/registry/DataFormat/specs/1.3/dataformat.1.3.html#_additional_dimensions_descriptor_block . I didn't implement either of those as they are out of scope for my needs (2D/3D uncompressed and compressed textures for real-time graphics usage.)

I suppose one could extract out the first 8 bytes of the BasicFormatDataDescriptor struct as a DataFormatDescriptorHeader and always parse it, then you conceptually want to have different types of the main body of the data format descriptor blocks based on what is in the header. The user doesn't need to see the header (necessarily? unless they want to be able to parse the data themselves because it's something that this crate doesn't know how to parse...) Hmm. Maybe the iterator should return something that can either be a BasicDataFormatDescriptor body, or something 'unknown' that contains the DataFormatDescriptorHeader and the body data to support external parsing? Rather than just erroring because this crate doesn't know what the data format descriptor is. I'd want to do it as an enum but I don't know if that is feasible as I think enum members need to be Sized and we won't know what the size of the custom data format descriptor blocks would be until they are parsed.

superdump replied:
Upon trying it, I don't like the enum approach as different information is needed from each variant and it feels weird to lump it all together. As such, I chose instead to try this:
```rust
pub struct DataFormatDescriptorHeader {
    pub vendor_id: u32,             //: 17;
    pub descriptor_type: u32,       //: 15;
    pub version_number: u32,        //: 16;
    pub descriptor_block_size: u32, //: 16;
}

pub struct DataFormatDescriptor<'data> {
    pub header: DataFormatDescriptorHeader,
    pub data: &'data [u8],
}

pub enum DataFormatDescriptorType {
    Basic, // vendor_id: 0, descriptor_type: 0, version_number: 2
    Unknown,
}

pub struct BasicDataFormatDescriptor<'data> {
    /// None means Unspecified or is an otherwise unknown value
    pub color_model: Option<ColorModel>, //: 8;
    /// None means Unspecified or is an otherwise unknown value
    pub color_primaries: Option<ColorPrimaries>, //: 8;
    /// None means Unspecified or is an otherwise unknown value
    pub transfer_function: Option<TransferFunction>, //: 8;
    pub flags: DataFormatFlags,     //: 8;
    pub texel_block_dimensions: [u32; 4], //: 8 x 4;
    pub bytes_planes: [u32; 8],     //: 8 x 8;
    sample_data: &'data [u8],
}
```

If there is not enough data to parse the header, I return None from the data format descriptors iterator. Otherwise I parse the header. Then if there is not enough data for the block then I return None. Otherwise I create a DataFormatDescriptor from the header and the byte range of the data. The DataFormatDescriptor implements a data_format_descriptor_type function that returns a DataFormatDescriptorType. I would kind of prefer to have a `try_into()` but yet again, not having std prevents that unless I define my own... which I guess I could do. I'm thinking that people can just try_into::<BasicDataFormatDescriptor>() and handle the error. What do you think of this?

superdump replied:
Alternatively I am back again to wanting to have constructors that return `Option<Self>` or `Result<Self, Something>`

Ralith replied:
Decoupling out the header, which is guaranteed to have a particular structure, and the `BasicDataFormatDescriptor`, which is not necessarily applicable, makes sense to me. Further, is the `descriptor_block_size` useful to the user, or is it redundant to the already suitably sliced `data` field of `DataFormatDescriptor`? Particularly if that field is omitted, we can then derive `Eq + PartialEq` for `DataFormatDescriptorHeader` and define an associated constant corresponding to the header that identifies a `BasicDataFormatDescriptor` which can be easily `match`ed against downstream, and gracefully extended with future descriptor types. This obviates the need for a `data_format_descriptor_type` method and the associated enum without compromising ergonomics.

We could further define pseudo-enums for vendor ID and descriptor type, but because there do not appear to exist any vendor-defined DFDs this is probably premature.

Given that there are a variety of possible present and future descriptor types, not all of which are necessarily known to this library, I think a fallible parsing constructor for `BasicDataFormatDescriptor` is justified. I suggest:

```rust
impl BasicDataFormatDescriptor<'data> {
    pub fn parse(data: &'data [u8]) -> Result<Self, MalformedDfd> { ... }
}

#[derive(Debug, Copy, Clone)]
struct MalformedDfd;
// typical error impls...
```

This does present users with the possibility of a successfully parsed file with malformed DFDs, which is unfortunate; however, because not all DFD types can be known in advance, this is de facto possible regardless as a file with unknown DFDs might be undetectably malformed, so we might as well embrace it. Perhaps @cwfitzgerald would like to weigh in here as well.

superdump replied:
`descriptor_block_size` is indeed redundant so I removed it, derived `PartialEq + Eq`, defined a constant for the header values for `BasicDataFormatDescriptor`, got rid of `data_format_descriptor_type()` and associated enum.

I definitely like using `parse() -> Result` as a fallible constructor and that leads me to think that the `bytes_to_u32(bytes: &[u8], offset: &mut usize) -> u32` helper that reads 4 bytes into a u32 should itself be fallible and return `Result<u32, ParseError>`, basically returning `ParseError::UnexpectedEnd` if it fails to read the 4 bytes.

Let me know what you think now.

Ralith replied:
I'm a fan of that pattern, since it makes it very hard to forget a bounds check.

Ralith commented on src/lib.rs:
```diff
@@ -162,3 +219,206 @@ impl LevelIndex {
+
+        let v = bytes_to_u32(bytes, &mut offset);
+        let vendor_id = shift_and_mask_lower(0, 17, v);
+        assert_eq!(vendor_id, 0); // Basic Data Format Descriptor requirement
+        let descriptor_type = shift_and_mask_lower(17, 15, v);
+        assert_eq!(descriptor_type, 0); // Basic Data Format Descriptor requirement
+
+        let v = bytes_to_u32(bytes, &mut offset);
+        let version_number = shift_and_mask_lower(0, 16, v);
+        assert_eq!(version_number, 2); // Basic Data Format Descriptor requirement
```
We must not panic on malformed data.

superdump replied:
Indeed, see my other comment about the data format descriptor blocks.

superdump committed:
Iterate on review feedback

Fallible parsing returning ParseError
Separate out DataFormatDescriptorHeader. Remove size member. Derive Eq +
PartialEq. Add constant for DFD header for Basic variant. Remove
data_format_descriptor_type method.

Ralith commented on src/lib.rs:
```diff
@@ -98,6 +97,65 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
+    pub fn data_format_descriptors(&self) -> ParseResult<impl Iterator<Item = DataFormatDescriptor>> {
+        let header = self.header();
+        let start = header.dfd_byte_offset as usize;
+        let length: u32 = u32::from_le_bytes(
+            self.input
+                .as_ref()
+                .get(start..start + 4)
+                .ok_or(ParseError::UnexpectedEnd)?
+                .try_into()
+                .unwrap(),
+        );
```
Should this use `bytes_to_u32`?

superdump replied:
Done.

Ralith commented on src/lib.rs:
```diff
@@ -98,6 +97,65 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
+        &self.input.as_ref()[start..end]
+    }
+
+    pub fn data_format_descriptors(&self) -> ParseResult<impl Iterator<Item = DataFormatDescriptor>> {
```
Since we can easily validate this consistently in advance, I'd prefer to do so in `new`, then have the public method be infallible.

superdump replied:
Done.

superdump committed:
Validate data format descriptor in Reader::new()

superdump committed:
Fix data format descriptor iterator data slice start

Ralith commented on src/lib.rs:
```diff
@@ -50,12 +50,22 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
+        // Check data format descriptor integrity
+        let mut offset = header.dfd_byte_offset as usize;
+        let length = bytes_to_u32(input.as_ref(), &mut offset)?;
+        if length != header.dfd_byte_length
```
`UnexpectedEnd` is a confusing error to return for what is actually an inconsistent/illegal field value. We could add a new error case, but given that the spec guarantees that these values are equal, I'd also be comfortable just omitting the check; worst-case scenario is that a malformed file yields garbage data, which is always possible regardless.

superdump replied:
In that case I guess I can just sanity check the fields used to slice the data to make sure they're in-bounds. As I use `dfd_byte_offset + dfd_byte_length` as then end, I could just check that the length of the data is >= that. Then returning `UnexpectedEnd` makes sense to me.

Ralith replied:
Yeah, sounds sensible.

Ralith commented on src/lib.rs:
```diff
@@ -105,47 +115,34 @@ impl<Data: AsRef<[u8]>> Reader<Data> {
 
     fn next(&mut self) -> Option<Self::Item> {
-        if self.offset > self.end - DataFormatDescriptorHeader::LENGTH {
+        if self.offset > self.data.len() - DataFormatDescriptorHeader::LENGTH {
```
Rather than tracking `offset` separately, could we just advance the start of `data`?

superdump replied:
Nice. Done.

Ralith reviewed (changes requested):
Thanks, this is looking good! Just a couple nits left.

superdump committed:
Only check if there is enough data for the data format descriptor block

superdump committed:
Re-slice &[u8] instead of maintaining a start offset

Ralith reviewed (approved):
LGTM, thanks for all your work!

superdump commented:
> LGTM, thanks for all your work!

Thanks for all the review! :) It’s for sure a better PR because of it.
